### DISCUSSION TITLE
What is the difference between `brew list --installed-on-request` and `brew leaves --installed-on-request`?

### DISCUSSION THREAD

**Discussion by nbro** [Homebrew/discussions#5332] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.20
ORIGIN: https://github.com/Homebrew/brew
HEAD: c2ed3327c605c3e738359c9807b8f4cd6fec09eb
Last commit: 2 days ago
Core tap HEAD: 909536f5dbfda87216933bb641a8e81fc8f62bae
Core tap last commit: 46 minutes ago
Core tap JSON: 01 May 13:55 UTC
Core cask tap HEAD: 63ffb145a8c06092692969cf9ddfa5d5b6c5954b
Core cask tap last commit: 2 hours ago
Core cask tap JSON: 01 May 13:55 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 13.1.6 build 1316
Git: 2.32.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 12.7.4-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 13.4.1
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have the following deprecated, cask taps tapped:
  caskroom/cask
Untap them with `brew untap`.

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
  python@3.8

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  glib
  numpy
  kubernetes-cli
  gnupg

Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

Warning: Your Xcode (13.4.1) is outdated.
Please update to Xcode 14.2 (or delete it).
Xcode can be updated from the App Store.
```

### Description of issue

I originally thought that to get the list of packages I explicitly asked brew to install I should use `brew leaves --installed-on-request`, but today I realised that `brew list` also has the flag `--installed-on-request`.

So, what is the difference between `brew list --installed-on-request` and `brew leaves --installed-on-request`?

If I run them, I get different outputs, so they must be doing something different, but what?

Also, if `brew list --installed-on-request` shows the packages that I explicitly installed, what's the purpose of `brew leaves`? In general, what's the purpose of this `brew leaves` command? Couldn't it just be merged with `brew list`? What do the "leaves" refer to?

**Comment by ZhongRuoyu** [comment, marked as answer]
[`brew leaves`](https://docs.brew.sh/Manpage#leaves---installed-on-request---installed-as-dependency) shows the formulae that are not dependencies of any other formula or cask. So, if you installed package A on request, but A is also a dependency of some other package B, then A will not show up in the output of `brew leaves --installed-on-request` (but it should be in the output of `brew list --installed-on-request`).

> I originally thought that to get the list of packages I explicitly asked brew to install I should use `brew leaves --installed-on-request`

`brew list --installed-on-request` should be what you want here.

**Reply by nbro** [reply]
@ZhongRuoyu Thanks for answer. But couldn't leaves be merged into list? Why have to separate commands? `brew list` could have a flag `--leaves`. Have you considered that? I'm asking because in my view I don't see why `leaves` should really be a separate command, but maybe I'm missing some context and, anyway, it doesn't hurt too much

**Reply by ZhongRuoyu** [reply]
I agree with you that `brew list --leaves` would look better, but `brew leaves` has existed for too long (Homebrew/legacy-homebrew@3d933c33bf9ecda31cc5063c2d6560a12f5d5687) so it's not going to be removed.

Perhaps coincidentally, this was also raised recently in https://github.com/Homebrew/brew/issues/17117#issuecomment-2067912599.
